### PR TITLE
BUG: Remove reset of seeding in the ML EM `fit` method of `Factor`

### LINKED ISSUE #7357: BUG: Fit method for Factor model using the EM algorithm resets the RNG seed
#### Describe the bug

Every time you call the `fit` method to estimate a `Factor` model using maximum likelihood (implemented via the EM algorithm), the code calls `np.random.seed` to reset the seeding. This is a problem because if, say, you want to perform a Monte Carlo experiment that involves fitting a Factor model as described, the seeding will be reset and will jeopardize the Monte Carlo experiment.

#### Code Sample, a copy-pastable example if possible

Below is an example that randomly creates data from a multi-factor model. I then estimate the factor model using maximum likelihood. I create a loop that generates what should be 10 different randomly generated data sets and estimate the model on each independent data set. As you can see, the 9 of the 10 outputs are the same because `statsmodels` is resetting the seed. The first output is different from the remaining 9 because I set the seed myself at the beginning of my code.

```python
import numpy as np
import pandas as pd
import statsmodels.api as sm

np.random.seed(100)
N = 10
T = 40

def fit_factor():
    mu = np.zeros((N,1))
    beta = np.random.multivariate_normal(np.zeros(2), np.eye(2), size=N)
    f = np.random.multivariate_normal(np.zeros(2), np.eye(2), size=T).T
    epsilon = np.random.multivariate_normal(np.zeros(N), np.eye(N), size=T).T

    x = mu + beta @ f + epsilon
    mu_hat = x.mean(axis=1)
    x_tilde = x.T - mu_hat

    factor = sm.Factor(endog=x_tilde, n_factor=2, method='ml')
    fitted_factor = factor.fit()

    beta_hat = fitted_factor.loadings
    return epsilon[0,0], x[0,0], f[0,0]

for i in range(10):
    print(fit_factor())
```

This code produces the following output. Each line should be different, as they should come from independent random draws.

```
(-1.7046512057609624, -4.009211534145269, 1.6189816606752596)
(0.06541577104039452, -1.262462114689552, 0.023364219186334523)
(0.06541577104039452, -1.262462114689552, 0.023364219186334523)
(0.06541577104039452, -1.262462114689552, 0.023364219186334523)
(0.06541577104039452, -1.262462114689552, 0.023364219186334523)
(0.06541577104039452, -1.262462114689552, 0.023364219186334523)
(0.06541577104039452, -1.262462114689552, 0.023364219186334523)
(0.06541577104039452, -1.262462114689552, 0.023364219186334523)
(0.06541577104039452, -1.262462114689552, 0.023364219186334523)
(0.06541577104039452, -1.262462114689552, 0.023364219186334523)
```

This behavior will arise in a simpler example, such as the one below, but in that example the fit method will raise a warning because the fitting algorithm won't converge since the data isn't generate according to a factor model.

```python
import numpy as np
import statsmodels.api as sm

np.random.seed(100)
N = 10
T = 40

def fit_factor():
    epsilon = np.random.multivariate_normal(np.zeros(N), np.eye(N), size=T).T
    factor = sm.Factor(endog=epsilon, n_factor=2, method='ml').fit()
    return epsilon[0,0]

for i in range(10):
    print(fit_factor())
```

#### Expected Output

Each line in the output of the above code snippet should be different, since they are independent random draws. 

#### Output of ``import statsmodels.api as sm; sm.show_versions()``

<details>

INSTALLED VERSIONS
------------------
Python: 3.7.10.final.0

statsmodels
===========

Installed: 0.12.2 (C:\Users\Jeremy\Miniconda3\lib\site-packages\statsmodels)

Required Dependencies
=====================

cython: Not installed
numpy: 1.20.1 (C:\Users\Jeremy\Miniconda3\lib\site-packages\numpy)
scipy: 1.6.0 (C:\Users\Jeremy\Miniconda3\lib\site-packages\scipy)
pandas: 1.2.2 (C:\Users\Jeremy\Miniconda3\lib\site-packages\pandas)
    dateutil: 2.8.1 (C:\Users\Jeremy\Miniconda3\lib\site-packages\dateutil)
patsy: 0.5.1 (C:\Users\Jeremy\Miniconda3\lib\site-packages\patsy)

Optional Dependencies
=====================

matplotlib: 3.3.4 (C:\Users\Jeremy\Miniconda3\lib\site-packages\matplotlib)
    backend: module://ipykernel.pylab.backend_inline 
cvxopt: Not installed
joblib: 1.0.1 (C:\Users\Jeremy\Miniconda3\lib\site-packages\joblib)

Developer Tools
================

IPython: 7.19.0 (C:\Users\Jeremy\Miniconda3\lib\site-packages\IPython)
    jinja2: 2.11.3 (C:\Users\Jeremy\Miniconda3\lib\site-packages\jinja2)
sphinx: 3.5.1 (C:\Users\Jeremy\Miniconda3\lib\site-packages\sphinx)
    pygments: 2.8.0 (C:\Users\Jeremy\Miniconda3\lib\site-packages\pygments)
pytest: 6.2.2 (C:\Users\Jeremy\Miniconda3\lib\site-packages\pytest)
virtualenv: Not installed

</details>

### PR DESCRIPTION
**Pull request by jmbejara**
- [x] closes #7357
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

I removed the line of code in the `_fit_ml_em` method of `multivariate.Factor` that resets the RNG seed. Doing so broke 2 tests cases, but these were only broken due to precision error. These are fixed by simply increasing the number of iterations used to estimate the models in the test cases.

<details>

**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in master and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify you changes are well formatted by running 
  ```
  git diff upstream/master -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it help 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

**Comment by bashtage** [comment]
I think a more deterministic fix is something like

```
rs = np.random.RadnomState(1234)
rs.normal(...)
```

which should be identical to the current behaviour without touching the singleton state.  A better optional lets a user pass a RandomState which is passed to fit em when not None, and if None the code above is used.

**Comment by jmbejara** [comment]
@bashtage As you suggested, I added the option to pass a `RandomState` instance that defaults to a generator with the same seed as the original implementation. I reverted the changes I made to the tests, since the code now behaves as it did before unless the user supplies a different generator. Hope this is ok. Let me know if I can do anything else! Thanks

File: statsmodels/multivariate/factor.py [hunk 1]
```diff
@@ -433,12 +433,15 @@ def nscore(par):
...
         """estimate Factor model using EM algorithm
         """
         # Starting values
-        np.random.seed(3427)
-        load = 0.1*np.random.normal(size=(self.k_endog, self.n_factor))
+        if random_state is None:
+            rs = np.random.RandomState(3427)
+        else:
+            rs = random_state
+        load = 0.1 * rs.normal(size=(self.k_endog, self.n_factor))
```
Comment (bashtage) [inline statsmodels/multivariate/factor.py:444]: Could you switch to standard_normal?  It is faster than normal.

**Comment by bashtage** [comment]
Looks good. Should hve a test (or modify and existing test) to make sure that the singleton RandomState is not changed by calling em.

File: statsmodels/multivariate/factor.py (hunk 1 above)
Comment (jmbejara) [reply statsmodels/multivariate/factor.py]: Yep! Just did it.

**Comment by jmbejara** [comment]
@bashtage Yep! Just added a short test for it based on the original issue.

File: statsmodels/multivariate/tests/test_ml_factor.py [hunk 2]
```diff
@@ -78,12 +78,33 @@ def test_exact_em():
...
+    This test arose from https://github.com/statsmodels/statsmodels/issues/7357
+    """
+
+    def fit_factor():
+        T = 10
+        epsilon = np.random.multivariate_normal(np.zeros(3), np.eye(3), size=T).T
+        factor = Factor(endog=epsilon, n_factor=2, method='ml').fit()
+        return epsilon[0,0]
+
+    with warnings.catch_warnings():
```
Comment (bashtage) [inline statsmodels/multivariate/tests/test_ml_factor.py:99]: I thinkyou need to indent 4 spaces starting here.

File: statsmodels/multivariate/tests/test_ml_factor.py [hunk 3]
```diff
@@ -78,12 +78,33 @@ def test_exact_em():
...
             c_e.flat[::c_e.shape[0]+1] += uniq_e
             assert_allclose(c_e, c, rtol=1e-4, atol=1e-4)
 
 
+def test_fit_ml_em_random_state():
+    """Ensure Factor._fit_ml_em doesn't change numpy's singleton random state
+
+    This test arose from https://github.com/statsmodels/statsmodels/issues/7357
+    """
+
```
Comment (bashtage) [inline statsmodels/multivariate/tests/test_ml_factor.py:92]: alternative would be `initial = np.random.get_state()`
then later
```
final = np.random.get_state()
```
and check that initial was equal to `final`.
 
But your method gets the point across.

Comment (jmbejara) [reply statsmodels/multivariate/tests/test_ml_factor.py:92]: I like that better too. Should have looked more into this. Just changed the test to do it this way.

**Pull request closed without merging**